presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -156,13 +206,48 @@ public PlanNode visitFilter(FilterNode node, RewriteContext<Constraint> context)
         {
             PlanNode rewrittenSource = context.rewrite(node.getSource(), new Constraint(Optional.empty(), Optional.of(node.getPredicate())));
             if (rewrittenSource != node.getSource()) {
-                if (rewrittenSource instanceof TopNRowNumberNode) {
+                if (rewrittenSource instanceof TopNRowNumberNode &&
+                        !shouldAddFilterNode((TopNRowNumberNode) rewrittenSource, node.getPredicate(), metadata, session, types)) {
                     return rewrittenSource;
                 }
                 return new FilterNode(idAllocator.getNextId(), rewrittenSource, node.getPredicate());
             }
             return context.defaultRewrite(node);
         }
+
+        // check whether we should add a FilterNode on top of TopNRowNumberNode
+        private static boolean shouldAddFilterNode(TopNRowNumberNode node, Expression predicate, Metadata metadata, Session session, Map<Symbol, Type> types)
+        {
```

Comment:
If we do not have a domain on the rowNumber, we could still have domains on other symbols. So when ``domain == null`` we need to add the filter node. 

The only condition we can skip the filter node is 
we have only one domain && 
it is a domain on the row number && 
this domain has exctly one range && 
it is upperbounded && 
it is not single value 

Instead of adding all these conditions, It seems a better idea to leave the FilterNode intact in all cases. I'll take this check out before merging the code.